### PR TITLE
mgr/dashboard: provisioned values is misleading in RBD image table

### PR DESCRIPTION
**Pull request by avanthakkar**
Fixes: https://tracker.ceph.com/issues/46617
Signed-off-by: Avan Thakkar <athakkar@redhat.com>

Adding hint in image table similar to the one in rbd-details.

Screenshot from 2021-11-09 22-42-18

## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [x] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by avanthakkar** [comment]
jenkins test make check

**Comment by avanthakkar** [comment]
jenkins test make check

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html [hunk 1]
```diff
@@ -29,6 +29,26 @@
        [innerHtml]="'Only available for RBD images with <strong>fast-diff</strong> enabled'"></div>
 </ng-template>
 
+<ng-template #provisionedNotAvailableTooltipTpl
+             let-row="row">
+  <span *ngIf="row.disk_usage === null; else provisioned"
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html:34]:
```suggestion
  <span *ngIf="!row.disk_usage; else provisioned"
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html [hunk 2]
```diff
@@ -29,6 +29,26 @@
...
+        [ngbTooltip]="usageNotAvailableTooltipTpl"
+        placement="top"
+        i18n>N/A</span>
+  <ng-template #provisioned
+               i18n>{{row.disk_usage | dimlessBinary}}</ng-template>
+</ng-template>
+
+<ng-template #totalProvisionedNotAvailableTooltipTpl
+             let-row="row">
+  <span *ngIf="row.total_disk_usage === null; else totalProvisioned"
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html:44]:
```suggestion
  <span *ngIf="!row.total_disk_usage; else totalProvisioned"
```

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html [hunk 3]
```diff
@@ -29,6 +29,26 @@
        [innerHtml]="'Only available for RBD images with <strong>fast-diff</strong> enabled'"></div>
 </ng-template>
 
+<ng-template #provisionedNotAvailableTooltipTpl
+             let-row="row">
+  <span *ngIf="row.disk_usage === null; else provisioned"
+        [ngbTooltip]="usageNotAvailableTooltipTpl"
+        placement="top"
+        i18n>N/A</span>
```
Comment (nizamial09) [inline src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html:37]: We are showing an `Unavailable` in the hosts table when orchestrator is not available for the gather facts field. I think its better to be consistent and show Unavailable over here too? @avanthakkar WDYT?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html (hunk 1 above)
Comment (epuertat) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html:34]: @nizamial09 what if `disk_usage == 0`?

**Review by epuertat — COMMENTED** [review]
I'm a lazy developer (and I'm becoming lazier every day), so I was wondering: instead of adding all this code to deal with this specific case, why not just modifying the behaviour of the `dimlessBinary` pipe to show an `N/A ` when the input is `null`? That behaviour wouldn't be inconsistent, right? If the input is not a number, you simply return "N/A" or "-" (which I think it's the default behaviour). Not only will that fix this, but every other potential case where this would apply, wouldn't it?

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html (hunk 3 above)
Comment (avanthakkar) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html:37]: I think that's for another reason, here I've shown `N/A` following what's shown in details section as well (value=`N/A`  and the similar text tooltip)

**Comment by avanthakkar** [comment]
> I'm a lazy developer (and I'm becoming lazier every day), so I was wondering: instead of adding all this code to deal with this specific case, why not just modifying the behaviour of the `dimlessBinary` pipe to show an `N/A ` when the input is `null`? That behaviour wouldn't be inconsistent, right? If the input is not a number, you simply return "N/A" or "-" (which I think it's the default behaviour). Not only will that fix this, but every other potential case where this would apply, wouldn't it?

But I'd still need tooltip template and for tht I'll need the condition for `null` checking, so not sure how to make this work here

**Comment by epuertat** [comment]
> > I'm a lazy developer (and I'm becoming lazier every day), so I was wondering: instead of adding all this code to deal with this specific case, why not just modifying the behaviour of the `dimlessBinary` pipe to show an `N/A ` when the input is `null`? That behaviour wouldn't be inconsistent, right? If the input is not a number, you simply return "N/A" or "-" (which I think it's the default behaviour). Not only will that fix this, but every other potential case where this would apply, wouldn't it?
> 
> But I'd still need tooltip template and for tht I'll need the condition for `null` checking, so not sure how to make this work here

* Given having tool tips in column headers seems a desirable functionality, why not adding a `tooltip` property to `CdTableColumn`, and providing the tooltip without the need of a custom cell template?
* The dimless-binary pipe simply calls the `FormatterService` service, and in case of `null` it just dumps a `-`. Are we sure we prefer "N/A" to that?

**Comment by avanthakkar** [comment]
> > > I'm a lazy developer (and I'm becoming lazier every day), so I was wondering: instead of adding all this code to deal with this specific case, why not just modifying the behaviour of the `dimlessBinary` pipe to show an `N/A ` when the input is `null`? That behaviour wouldn't be inconsistent, right? If the input is not a number, you simply return "N/A" or "-" (which I think it's the default behaviour). Not only will that fix this, but every other potential case where this would apply, wouldn't it?
> > 
> > 
> > But I'd still need tooltip template and for tht I'll need the condition for `null` checking, so not sure how to make this work here
> 
>     * Given having tool tips in column headers seems a desirable functionality, why not adding a `tooltip` property to `CdTableColumn`, and providing the tooltip without the need of a custom cell template?
> 
But we need to somehow access the cell value and this is like  cellTemplate having a property not Table Column itself I think. I'm not following this suggestion exactly.
>     * The dimless-binary pipe simply calls the `FormatterService` service, and in case of `null` it just dumps a `-`. Are we sure we prefer "N/A" to that?

This isn't  a problem I think , null condition is already handled (w/o dimlessBinary) so it'll have N/A text and value similar to what we show in rbd-image details section as well.

File: src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html (hunk 1 above)
Comment (nizamial09) [reply src/pybind/mgr/dashboard/frontend/src/app/ceph/block/rbd-list/rbd-list.component.html:34]: Hmm you are right. then `null` should be the one! Thanks @epuertat

**Review by nizamial09 — APPROVED** [review]

**Comment by epuertat** [comment]
> But we need to somehow access the cell value and this is like cellTemplate having a property not Table Column itself I think. I'm not following this suggestion exactly.

My suggestion was that instead of creating another template for dealing with 2 different values (null, and not null) you may use a pipe to present 2 different outputs (we even have a `notAvailable` pipe but that would need to be extended to cover null value).

**Review by epuertat — APPROVED** [review]

**Pull request merged**